src/index.ts

```diff
@@ -1,28 +1,84 @@
-import * as ended from 'is-stream-ended';
 import {Transform} from 'stream';
 
+// tslint:disable-next-line no-any
+export type ArrayItem = any;
+
 /**
  * Push an array of items into a Transform stream.
- * @param array The array you want to push to the stream.
- * @param stream The Transform stream into which array items are pushed.
+ * @param getArrayFn The array you want to push to the stream.
  */
-export async function split(array: Array<{}>, stream: Transform) {
-  return new Promise<boolean>((resolve, reject) => {
-    const arr = [].slice.call(array);
-    function loopyloop() {
-      // Ensure the stream wasn't closed by the consumer.
-      const isEnded = ended(stream);
-      // Ensure al items from the array haven't been pushed.
-      const cont = !isEnded && arr.length > 0;
-      if (cont) {
-        stream.push(arr.shift());
-        // For large arrays, use setImmediate to ensure other microtasks
-        // and I/O operations have a chance to execute.
-        setImmediate(loopyloop);
-      } else {
-        resolve(isEnded);
-      }
-    }
-    loopyloop();
-  });
-}
+export class SplitArrayStream extends Transform {
+  private _ended: boolean;
+  private _getArrayFn?: Function;
+
+  constructor(getArrayFn?: Function|ArrayItem[]) {
+    super({objectMode: true});
+
+    this._ended = false;
+
+    // When the input is an array, the user just wants it split up and emitted
+    // as data events.
+    if (Array.isArray(getArrayFn)) {
+      const array = [].slice.call(getArrayFn);
+      array.push(null);
+      this._getArrayFn = () => Promise.resolve(array);
+    }
+
+    // When the input is a function, the user wants that function called each
+    // time the destination stream is ready for more data.
+    if (typeof getArrayFn === 'function') {
+      this._getArrayFn = getArrayFn;
+    }
+
+    if (typeof getArrayFn !== 'undefined') {
+      this._read = this._readFromFn.bind(this);
+    }
+  }
+
+  // tslint:disable-next-line no-any
+  end(...args: any[]) {
+    this._ended = true;
+    return super.end(...args);
+  }
+
+  async _readFromFn() {
+    let consumerStreamReady = true;
+
+    let arrayValue = await this._getArrayFn!();
+
+    if (!Array.isArray(arrayValue)) {
+      arrayValue = [arrayValue];
+    }
+
+    const array = [].slice.call(arrayValue);
+
+    while (!this._ended && consumerStreamReady && array.length > 0) {
+      consumerStreamReady = this.push(array.shift());
```

Comment:
I found in my benchmarking that calling `shift()` on a large array can be very slow. I opted for a `for ... of` with a conditional `break` statement instead.